spring-cloud-stream/src/main/java/org/springframework/cloud/stream/binding/MessageConverterConfigurer.java

```diff
@@ -185,116 +193,79 @@ public int selectPartition(Object key, int partitionCount) {
 
 	}
 
-	private final class ContentTypeConvertingInterceptor extends ChannelInterceptorAdapter {
-
-		private final String contentType;
+	private final class ContentTypeConvertingInterceptor
+			extends ChannelInterceptorAdapter {
 
 		private final MimeType mimeType;
 
 		private final boolean input;
 
-		private final Class<?> klazz;
-
 		private final MessageConverter messageConverter;
 
 		private final boolean provideHint;
 
 		private ContentTypeConvertingInterceptor(String contentType, boolean input) {
-			this.contentType = contentType;
 			this.mimeType = MessageConverterUtils.getMimeType(contentType);
 			this.input = input;
-			if (MessageConverterUtils.X_JAVA_OBJECT.includes(this.mimeType)) {
-				this.klazz = MessageConverterUtils
-						.getJavaTypeForJavaObjectContentType(this.mimeType);
-			}
-			else if (this.mimeType.equals(MessageConverterUtils.X_SPRING_TUPLE)) {
-				this.klazz = Tuple.class;
-			}
-			else if (this.mimeType.getType().equals("text") || this.mimeType.getSubtype().equals(
-					"json") || this.mimeType.getSubtype().equals("xml")) {
-				this.klazz = String.class;
-			}
-			else {
-				this.klazz = byte[].class;
-			}
+
 			this.messageConverter = MessageConverterConfigurer.this.compositeMessageConverterFactory
-					.getMessageConverterForType(this.mimeType);
+					.getMessageConverterForAllRegistered();
 			this.provideHint = this.messageConverter instanceof AbstractMessageConverter;
 		}
 
 		@Override
 		public Message<?> preSend(Message<?> message, MessageChannel channel) {
+			// bypass conversion for ErrorMessges
+			if (message instanceof ErrorMessage) {
+				return message;
+			}
+			// bypass conversion for raw bytes
+			if (message.getPayload() instanceof byte[]) {
+				return MessageConverterConfigurer.this.messageBuilderFactory
```

Comment:
Well, we had a discussion that we are going to use here `MutableMessageBuilder` for better performance.
Are you going to do that in the separate PR as well?

Reply:
I thought about leaving that for a separate PR as well @artembilan since it's more of a performance optimization. To be fair I need to understand more how would it work, but I'm not forgetting about it

Reply:
It's here: https://github.com/spring-cloud/spring-cloud-stream/issues/1058